### PR TITLE
Implement family functionality for scalars

### PR DESCRIPTION
**Pull request by davschneller**
Right now, scalars need to be globally unique. Maybe we can change that again (in SeisSol, there was only one collision).

### CODE REVIEW

File: yateto/codegen/code.py [hunk 1]
```diff
@@ -132,6 +132,9 @@ def If(self, argument):
       
   def For(self, argument):
     return Block(self, 'for ({})'.format(argument))
+
+  def ForRange(self, variable, rang):
```
Comment (krenzland) [inline yateto/codegen/code.py:136]: rang -> range

File: yateto/codegen/visitor.py [hunk 2]
```diff
@@ -572,8 +601,22 @@ def __init__(self, arch, tensors):
         self._collect[baseName][group] = tensor
       else:
         assert self._collect[baseName][group] == tensor
+    for tensor in scalars:
```
Comment (krenzland) [inline yateto/codegen/visitor.py:604]: for scalar in scalars:

File: yateto/codegen/visitor.py [hunk 3]
```diff
@@ -585,6 +628,7 @@ def _tensorViewGenerator(self, memoryLayout):
   def iterate_collect(self):
     cur_namespace = ''
     cur_dict = collections.OrderedDict()
+    cur_dictScalar = collections.OrderedDict()
```
Comment (krenzland) [inline yateto/codegen/visitor.py:631]: cur_dict_scalar or curDictScalar

File: yateto/codegen/visitor.py [hunk 4]
```diff
@@ -599,7 +643,26 @@ def iterate_collect(self):
       cur_dict[base_name, base_name_without_ns] = tensors
     # Don't forget last namespace
     yield cur_namespace, cur_dict
-    
+
+  def iterate_collect_scalar(self):
+    cur_namespace = ''
+    cur_dict = collections.OrderedDict()
+    cur_dictScalar = collections.OrderedDict()
+    for base_name, tensors in self._scalarCollect.items():
```
Comment (krenzland) [inline yateto/codegen/visitor.py:651]: base_name, scalar

File: yateto/codegen/visitor.py [hunk 5]
```diff
@@ -599,7 +643,26 @@ def iterate_collect(self):
...
+      splitted = base_name.rsplit('::', 1)
```
Comment (krenzland) [inline yateto/codegen/visitor.py:652]: split

File: yateto/codegen/visitor.py [hunk 6]
```diff
@@ -627,6 +690,27 @@ def generateTensorsH(self, header):
...
+            if len(groupSize) > 0:
+              with header.Function(self.INDEX_FUN_NAME, typedArgs, returnType):
+                header('return {};'.format(indexFun(groupSizeToStride(groupSize))))
+            if len(groupSize) > 0:
+              header('template<typename T>')
+              with header.Struct(self.CONTAINER_CLASS_NAME):
+                header('T {}[{}];'.format(self.CONTAINER_DATA_NAME, reduce(operator.mul, groupSize)))
+                with header.Function(self.CONTAINER_CLASS_NAME, '', ''):
+                  pass
+                # header('{}() : {}{{}} {{}}'.format(self.CONTAINER_CLASS_NAME, self.CONTAINER_DATA_NAME))
```
Comment (krenzland) [inline yateto/codegen/visitor.py:709]: remove

File: yateto/type.py [hunk 7]
```diff
@@ -12,17 +12,65 @@ def isValidName(cls, name):
   def name(self):
     return self._name
 
+class IdentifiedType(AbstractType):
+  BASE_NAME = r'[a-zA-Z]\w*'
+  GROUP_INDEX = r'(0|[1-9]\d*)'
+  GROUP_INDICES = r'\(({0}(,{0})*)\)'.format(GROUP_INDEX)
+  VALID_NAME = r'^{}({})?$'.format(BASE_NAME, GROUP_INDICES)
+
+  # TODO:(David) superclass Scalar and Tensor
```
Comment (krenzland) [inline yateto/type.py:21]: !

**Review by krenzland — CHANGES_REQUESTED** [review]
Some small things. Looks okay but the overall design is very confusing (but that's IMO not your fault :D)

File: yateto/codegen/visitor.py (hunk 3 above)
Comment (davschneller) [reply yateto/codegen/visitor.py]: I forgot to remove that one

File: yateto/codegen/visitor.py (hunk 5 above)
Comment (davschneller) [reply yateto/codegen/visitor.py:653]: `rsplit` should be correct here: e.g. `a::b::c::krnl` should result in the kernel name `krnl` and the namespace `a::b::c`

File: yateto/type.py (hunk 7 above)
Comment (davschneller) [reply yateto/type.py]: Should be done now.

File: yateto/codegen/visitor.py (hunk 5 above)
Comment (krenzland) [reply yateto/codegen/visitor.py:653]: Sorry, what I meant: 'splitted' isn't a word...

Comment (davschneller) [reply yateto/codegen/visitor.py]: Ok, renamed it.

Comment (davschneller) [reply yateto/codegen/visitor.py]: (even though there is still blatant code duplication w.r.t. the method above it)

**Review by krenzland — APPROVED** [review]

**Pull request merged**